File: 2026/AHC060/a06.py
import os
from typing import Tuple, List
import random
from typing import Optional
import time

MyPC = os.path.basename(__file__) != "Main.py"
if MyPC:
  from icecream import ic
  ic.disable()
else:
  def ic(*args):
    return None

ic.enable() if MyPC else None

start_time = time.time()

random.seed(1317)

# 定数
N = 100
K = 10
T = 10000

# 方針03用の定数
P = 0.02  # ランダムに選ぶ確率

TIME_LIMIT = 1.9  # 秒

def calc_score(outputs: List[int]) -> int:
  """出力列をシミュレーションしてスコアを返す。

  outputs:
    - v (0 <= v < N): 行動1で頂点vへ移動（木なら収穫、店なら納品）
    - -1: 行動2（現在位置の木を W -> R に変更）

  注意:
    - この関数は隣接リストを受け取らないため、辺の妥当性チェックは行わない。
    - 不正な出力列（範囲外、戻り禁止違反、赤化不可など）の場合は大きな負値を返す。
  """

  red = [False] * N  # 木の赤化状態（ショップは無視）
  inventories = [set() for _ in range(K)]

  pos = 0
  prev_move_from: Optional[int] = None  # 直前の行動1の移動元
  cone_chars: List[str] = []

  for out in outputs:
    if out == -1:
      # 行動2: 現在位置が白い木のときのみ可
      if not (K <= pos < N) or red[pos]:
        return -10**18
      red[pos] = True
      continue

    v = out
    if not (0 <= v < N):
      return -10**18
    if prev_move_from is not None and v == prev_move_from:
      return -10**18

    prev_move_from = pos
    pos = v

    if pos < K:
      inventories[pos].add("".join(cone_chars))
      cone_chars.clear()
    else:
      cone_chars.append('R' if red[pos] else 'W')

  return sum(len(inventories[i]) for i in range(K))

def main():
  _, M, _, _ = map(int, input().split())  # N: 頂点の数, M: 辺の数, K: アイス屋の数, T: ターン数  M以外は定数なので無視
  adj_list: List[List[int]] = [[] for _ in range(N + 1)]  # 隣接リストの初期化
  for _ in range(M):
    a, b = map(int, input().split())
    adj_list[a].append(b)
    adj_list[b].append(a)
  
  # ~~この後に座標が入力されるが、使用しないので無視~~
  # やっぱりアイス屋の位置は必要なのでK(=10)行だけ読み込む
  ice_shops: List[Tuple[int, int]] = []
  for _ in range(K):
    x, y = map(int, input().split())
    ice_shops.append((x, y))

  max_score = -1
  best_outputs: List[int] = []

  while time.time() < start_time + TIME_LIMIT:
    is_red = [False] * N  # 各頂点が赤くなっているかどうかのフラグ
    # 完全ランダムに移動
    outputs: List[int] = []
    current_pos = 0  # 最初は頂点0にいるものとする
    prev_pos = -1  # 直前の位置（最初は存在しないので-1）
    step_count = 0  # 移動回数カウント
    while step_count < T:
      # 次の移動先をランダムに決定
      candidates = adj_list[current_pos]  # 浅いコピー
      # 直前の位置に戻るのは避ける。常に候補が2つ以上あるので問題ないはず
      candidates = [v for v in candidates if v != prev_pos]
      next_pos = random.choice(candidates)
      outputs.append(next_pos)
      prev_pos = current_pos
      current_pos = next_pos
      step_count += 1
      # 頂点番号がK以上の場合、確率で赤くする
      if current_pos >= K and not is_red[current_pos]:
        if random.random() < P:
          is_red[current_pos] = True
          outputs.append(-1)
          step_count += 1

    score = calc_score(outputs)
    if score > max_score:
      max_score = score
      best_outputs = outputs[:]
      ic(score)
  ic(max_score)

  # 出力
  for v in best_outputs:
    print(v)


if __name__ == "__main__":
  main()

File: 2026/AHC060/a06.cpp
#include <bits/stdc++.h>
using namespace std;

// 定数（問題条件で固定）
static const int N = 100;
static const int K = 10;
static const int T = 10000;

// 方針03用
static const double P = 0.02;       // 木を赤化する確率
static const double TIME_LIMIT = 1.99; // 秒

// シミュレーションしてスコア計算（Pythonの calc_score 相当）
long long calc_score(const vector<int>& outputs) {
    vector<char> red(N, 0); // 木の赤化状態（ショップは無視）
    vector<unordered_set<string>> inventories(K);

    int pos = 0;
    bool has_prev_move_from = false;
    int prev_move_from = -1;
    string cone;  // 手元のコーン（味列）

    for (int out : outputs) {
        if (out == -1) {
            // 行動2: 現在位置が白い木のときのみ可
            if (!(K <= pos && pos < N) || red[pos]) return -(long long)1e18;
            red[pos] = 1;
            continue;
        }

        int v = out;
        if (!(0 <= v && v < N)) return -(long long)1e18;
        if (has_prev_move_from && v == prev_move_from) return -(long long)1e18;

        prev_move_from = pos;
        has_prev_move_from = true;
        pos = v;

        if (pos < K) {
            inventories[pos].insert(cone);
            cone.clear();
        } else {
            cone.push_back(red[pos] ? 'R' : 'W');
        }
    }

    long long sum = 0;
    for (int i = 0; i < K; i++) sum += (long long)inventories[i].size();
    return sum;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int Nin, M, Kin, Tin;
    cin >> Nin >> M >> Kin >> Tin; // N,K,Tは固定だが入力として来るので読む（無視してOK）

    vector<vector<int>> adj(N);
    for (int i = 0; i < M; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    // 座標入力（この解では使わない）
    // Python版はK行だけ読むが、どちらでもOK。ここでは全部読み捨てしておく。
    for (int i = 0; i < N; i++) {
        int x, y;
        cin >> x >> y;
    }

    // 乱数
    mt19937 rng(1317);
    uniform_real_distribution<double> real01(0.0, 1.0);

    long long max_score = -1;
    vector<int> best_outputs;

    using Clock = chrono::steady_clock;
    auto start = Clock::now();

    auto elapsed_sec = [&]() -> double {
        return chrono::duration<double>(Clock::now() - start).count();
    };

    while (elapsed_sec() < TIME_LIMIT) {
        vector<char> is_red(N, 0);
        vector<int> outputs;
        outputs.reserve(T);

        int current_pos = 0;
        int prev_pos = -1;
        int step_count = 0;

        while (step_count < T) {
            // 次の移動先候補（戻り禁止: prev_pos を除外）
            const auto& nbr = adj[current_pos];
            vector<int> candidates;
            candidates.reserve(nbr.size());
            for (int v : nbr) if (v != prev_pos) candidates.push_back(v);

            // 2-辺連結 & 関節点なし想定なので候補は必ず残る
            uniform_int_distribution<int> pick(0, (int)candidates.size() - 1);
            int next_pos = candidates[pick(rng)];

            outputs.push_back(next_pos);
            prev_pos = current_pos;
            current_pos = next_pos;
            step_count++;

            // 木なら確率で赤化（ただしT手を超えない範囲で）
            if (current_pos >= K && !is_red[current_pos]) {
                if (real01(rng) < P && step_count < T) {
                    is_red[current_pos] = 1;
                    outputs.push_back(-1);
                    step_count++;
                }
            }
        }

        long long score = calc_score(outputs);
        if (score > max_score) {
            max_score = score;
            best_outputs = outputs;
        }
    }

    // 出力
    for (int v : best_outputs) {
        cout << v << '\n';
    }
    return 0;
}
